DataFormats/Portable/interface/PortableHostCollectionReadRules.h

```diff
@@ -0,0 +1,84 @@
+void readPortableHostCollection(char *target, TVirtualObject *from_buffer) {
+  // extract the actual types
+  using Collection = T;
+  using Layout = typename Collection::Layout;
+
+  // valid only for PortableHostCollection<T>
+  static_assert(std::is_same_v<Collection, PortableHostCollection<Layout>>);
+
+  // proxy for the object being read from file
+  struct OnFile {
+    Layout &layout_;
+  };
+
+  // address in memory of the buffer containing the object being read from file
+  char *address = static_cast<char *>(from_buffer->GetObject());
+  // offset of the "layout_" data member
+  static ptrdiff_t layout_offset = from_buffer->GetClass()->GetDataMemberOffset("layout_");
+  // reference to the Layout object being read from file
+  OnFile onfile = {*(Layout *)(address + layout_offset)};
+
+  // pointer to the Collection object being constructed in memory
+  Collection *newObj = (Collection *)target;
+
+  // move the data from the on-file layout to the newly constructed object
+  Collection::ROOTReadStreamer(newObj, onfile.layout_);
+}
+
+/* Set the read rules for PortableHostCollection<T>.
+ * This is called only once, when the dictionary is loaded.
+ */
+template <typename T>
+static std::vector<::ROOT::Internal::TSchemaHelper> build_PortableHostCollection_read_rules(std::string const &type) {
+  // build the read rules
+  std::vector<::ROOT::Internal::TSchemaHelper> readrules(1);
+  ::ROOT::Internal::TSchemaHelper *rule = &readrules[0];
+  rule->fTarget = "buffer_,layout_,view_";
+  rule->fSourceClass = type;
+  rule->fSource = type + "::Layout layout_;";
+  rule->fCode = type + "::ROOTReadStreamer(newObj, onfile.layout_)";
+  rule->fVersion = "[1-]";
+  rule->fChecksum = "";
+  rule->fInclude = "";
+  rule->fEmbed = false;
+  rule->fFunctionPtr = (void *)TFunc2void(readPortableHostCollection<T>);
+  rule->fAttributes = "";
+
+  // set the read rules
+  //::ROOT::GenerateInitInstance((T const *)nullptr)->SetReadRules(readrules);
+
+  return readrules;
+}
+
+#define SET_PORTABLEHOSTCOLLECTION_READ_RULES(COLLECTION)                                                  \
+  namespace ROOT {                                                                                         \
+    /* forward declaration */                                                                              \
+    TGenericClassInfo *GenerateInitInstance(COLLECTION const *);                                           \
+  }                                                                                                        \
+                                                                                                           \
+  static bool EDM_CONCATENATE(set_PortableHostCollection_read_rules_done_at_, __LINE__) =                  \
+      (::ROOT::GenerateInitInstance((COLLECTION const *)nullptr)                                           \
+           ->SetReadRules(build_PortableHostCollection_read_rules<COLLECTION>(EDM_STRINGIZE(COLLECTION))), \
```

Comment:
Using the demangled name (_e.g._ `boost::core::demangle(typeid(COLLECTION).name())`) instead of the macro-based stringification does not work.
I assume it is because it fully expands the `typedef` (which expands to a template instantiation with various arguments) though I'm not sure _why_ that does not work.